design/20220118.certificate-issuance-exponential-backoff.md

```diff
@@ -0,0 +1,144 @@
+# Exponential backoff for certificate issuance
+
+## Summary
+
+Implement a way to apply exponential backoff when retrying failed certificate
+issuances. Issuance in this context refers to the period of time during which
+the `Issuing` condition on a `Certificate` is set to true and for which a new
+set of issuance-specific resources (`CertificateRequest`s, `Order`s, `Challenge`s etc) is
+created.
+
+## Motivation
+
+Currently failed issuances are retried once an hour without a backoff or time limit. This means that 1) continuous failures in large installations can overwhelm external services 2) in case of some external problems (i.e DNS) that cause issuances to fail and last for longer than a couple of hours, rate limits (see [Let'sEncrypt rate limts](https://letsencrypt.org/docs/rate-limits/)) can be easily hit.
+
+## Goals
+
+- Ensure that retrying failed issuances does not overwhelm external services and is less likely to cause issuance being rate-limited by backing off with an exponentially increasing time period. 
```

Comment:
a bit of a nit: I think it may be better to change "overwhelm external services" to more like "not use up all of a users quota" or something similar - I imagine this *isn't* the feature that prevents a DoS (i.e. huge numbers of HTTP requests) - rather it slows down the rate at which we start a whole new issuance flow.

I only call it out because there's other work going on that *does* aim to prevent overwhelming external services too - and this serves a subtly different (but equally important) area

Reply:
> a bit of a nit: I think it may be better to change "overwhelm external services" to more like "not use up all of a users quota" or something similar - I imagine this isn't the feature that prevents a DoS (i.e. huge numbers of HTTP requests) - rather it slows down the rate at which we start a whole new issuance flow.

This is a point where I sometimes struggle with imagining how these things are likely to interact in large scale deployments. For a small number of certs from user's perspective retrying every hour should be ok and the problem would likely be only exceeding a quota as you say.
But for a very large installation where something happens such that all cert issuances are suddenly failing and are now all being retried every hour, would the number of HTTP requests also not become an issue?
Similarly, assuming point of view of an external service (i.e Let'sEncrypt) that has a large number of cert-manager using clients- this should reduce the number of calls they receive _overall_ that are less likely to succeed